pennylane/ops/op_math/controlled_decompositions.py

```diff
@@ -15,10 +15,40 @@
 This submodule defines functions to decompose controlled operations
 """
 
+import numpy as np
+import numpy.linalg as npl
 import pennylane as qml
 from pennylane.operation import Operator
 from pennylane.wires import Wires
 
+def _matrix_adjoint(matrix: np.ndarray):
+    return np.transpose(np.conj(matrix))
+
+def _param_su2(ar: float, ai: float, br: float, bi: float):
+    """
+    Create a matrix in the SU(2) form from complex parameters a, b.
+    The resulting matrix is not guaranteed to be in SU(2), unless |a|^2 + |b|^2 = 1.
+    """
+    return np.array([[complex(ar, ai), complex(-br, bi)],
+                     [complex(br, bi), complex(ar, -ai)]])
+
+def _bisect_compute_a(u: np.ndarray):
+    """
+    Given the U matrix, compute the A matrix such that
+    At x A x At x A x = U
+    where At is the adjoint of A
+    and x is the Pauli X matrix.
+    """
+    x = np.real(u[0,1])
+    z = u[1,1]
+    zr = np.real(z)
+    zi = np.imag(z)
+    ar = np.sqrt((np.sqrt((zr+1)/2)+1)/2)
+    mul = 1/(2*np.sqrt((zr+1)*(np.sqrt((zr+1)/2)+1)))
+    ai = zi*mul
+    br = x*mul
+    bi = 0
+    return _param_su2(ar,ai,br,bi)
```

Comment:
Fine for now, but just a heads up we will want this code to be differentiable with all the different interfaces. To do so, we can use the `qml.math` module to dispatch to the relevant function in all the different interfaces. For example, `qml.math.real`, `qml.math.sqrt`, etc.

We can type hint arbitrary data from any interface with `from pennylane.typing import TensorLike`.

Reply:
It's been changed to use `math` instead of `np` now, all good?